designs/0005-static-matrices.md

```diff
@@ -0,0 +1,95 @@
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+At the language level, `static_matrix` can replace a `matrix` type if subsets of the matrix are never assigned to. That makes code such as the below illegal.
+
+```stan
+static_matrix[N, M] A = // Fill in A...
+A[10, 10] = 5;
+```
+
+Besides that they can be thought of as standard matrix.
+
+At the Stan Math level a `static_matrix` is a `var_type<Eigen::Matrix<double, -1, -1>>` with an underlying pointer to a `vari_type<Eigen::Matrix<double, -1, -1>>`*. This means that accessors for the value and adjoint of `var_type` objects can access contiguous chunks of each part of the `var_type`. Any function that accept a `var_type<T>` will support static matrices.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This feature will be added in the following stages.
+
+1. Replace `var` and `vari` with templated types. See the [example branch](https://github.com/stan-dev/math/compare/feature/var-template) here to see how this will be done. `var` and `vari` have been changed to `var_type<T>` and `vari_type<T>` with aliases for the current methods as `using var = var_type<double>` and `using vari = vari_type<double>`. These aliases allow for full backwards compatibility with existing upstream dependencies.
+
+2. Make the stack allocator conform to accept `vari_type<T>` objects.
+
+3. Additional PRs for the current arithmetic operators for `var` to accept `vari_type<T>` object types.
+
+4. Sort out and add any additional specialized methods for static matrix types such as `cholesky_decompose`.
+
+5. Add the `static_matrix` and `static_vector` types to the Stan language along with supported signatures.
+
+6. Support detection and substitution of `matrix` types for `static_matrix` types.
+
+7. ???
+
+8. Done!
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+More templates can be confusing and will lead to larger compile time. The confusion of templates can be mitigated by proper duty of care with documentation and guides for the Stan Math type system. Larger compile times can't be avoided with this implementation, though other forthcoming proposals can allow us to monitor increases in compilation times.
+
+With `sparse_matrix` and `complex_matrix` this will now add _another_ `*_matrix` type proposal to the language. There's no way to get around this in the Stan language currently, though some small side discussions exist to support attributes on types such as
+
+```stan
+@(sparse_type, static_type) // can't have static here because of C++ keyword
```

Comment:
> // can't have static here because of C++ keyword

Can you explain why not?

Reply:
That won't get in the way of our code---we just don't want conflicting things code generated.

Reply:
Oh my b I thought any C++ reserved keywords are banned. Not sure how we currently look for them though it may be good to avoid them so that the compiler doesn't have to check for "No C++ reserved keywords except for in these specific places"

Reply:
I like `const` more than `static`. `const` just implies you shouldn't be assigning. `static` is more like defining the scope of a variable (in Stan, it seems like `static` data would be the stuff that lives for the whole length of a chain, not just a gradient evaluation).

Reply:
Oh whoops I see this has already been said my bad.

Reply:
1.  We already use `for` and `if` and `int` and `while` and `return` and `continue`.  It's really just identifiers.  That's the only place *users* can define variables and we don't want users defining variables that overlap with C++ or Stan keywords.

2.  I agree that `const` means you shouldn't be assigning.  But that's not what's going on here.  We can assign, just not to individual elements.
    ```
    static_matrix a = b * c;
    a = a * d;  // OK to reassign
    a[1, 2] = 3;  // ILLEGAL to assign to elements
    ```
    Reassigning still maintains memory locality.  Assigning to elements does not.